sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -38,95 +260,208 @@ static kmac_error_t wait_status_bit(uint32_t bit_position, uint32_t bit_value) {
 }
 
 OT_WARN_UNUSED_RESULT
-kmac_error_t kmac_init(kmac_operation_t operation) {
+static kmac_error_t write_left_encode(const size_t len, uint8_t *encoded_len,
+                                      size_t *write_ctr) {
+  if (encoded_len == NULL || write_ctr == NULL || len > kKmacPrefixMaxSize) {
+    return kKmacArgsError;
+  }
+
+  size_t bit_len = len * 8;
+  if (bit_len < 256) {
+    // Need 1 byte to encode `len`
+    *encoded_len++ = 0x01;
+    *encoded_len++ = bit_len;
+    *write_ctr += 2;
+  } else {
+    // Need 2 bytes to encode `len`
+    *encoded_len++ = 0x02;
+    *encoded_len++ = bit_len / 256;
+    *encoded_len++ = bit_len % 256;
+    *write_ctr += 3;
+  }
+  // `bit_len` is upper bounded by kKmacPrefixSize * 8 = 288, so we don't need
+  // to implement larger encoding cases
+
+  return kKmacOk;
+}
+
+/*
+ * Set prefix registers.
+ */
+OT_WARN_UNUSED_RESULT
+static kmac_error_t kmac_set_prefix_regs(const uint8_t *func_name,
+                                         const size_t func_name_len,
+                                         const uint8_t *cust_str,
+                                         const size_t cust_str_len) {
+  // static_assert(kKmacPrefixMaxSize < 8192, ");
+
+  uint32_t prefix_buffer[kKmacPrefixRegCount] = {0x0};
+  uint8_t *prefix_buf_ptr = (uint8_t *)prefix_buffer;
+
+  if ((cust_str_len && cust_str == NULL) ||
+      (func_name_len && func_name == NULL)) {
+    return kKmacArgsError;
+  } else if (func_name_len + cust_str_len > kKmacPrefixMaxSize) {
+    return kKmacArgsError;
+  }
+
+  size_t func_name_encoded_len;
+  size_t cust_str_encoded_len;
+  size_t i = 0;
+  size_t write_cnt = 0;
   kmac_error_t err;
-  uint32_t cfg_reg;
-  uint32_t common_cfg_reg = KMAC_CFG_SHADOWED_REG_RESVAL;
 
-  // This probably needs to be set once per device boot rather than setting it
-  // at every KMAC invocation.
-  common_cfg_reg = bitfield_field32_write(
-      common_cfg_reg, KMAC_CFG_SHADOWED_ENTROPY_MODE_FIELD,
-      KMAC_CFG_SHADOWED_ENTROPY_MODE_VALUE_EDN_MODE);
+  // write left_encode of `func_name_len` and increment `write_cnt`
+  err = write_left_encode(func_name_len, prefix_buf_ptr, &write_cnt);
+  if (err != kKmacOk) {
+    return err;
+  }
+  // copy func_name
+  for (size_t i = 0; i < func_name_len; ++i) {
+    prefix_buf_ptr[write_cnt + i] = func_name[i];
+  }
+  write_cnt += func_name_len;
 
-  err = wait_status_bit(KMAC_STATUS_SHA3_IDLE_BIT, 1);
+  // write left_encode of cust_str_len
+  err = write_left_encode(cust_str_len, prefix_buf_ptr + write_cnt, &write_cnt);
   if (err != kKmacOk) {
     return err;
   }
+  // copy cust_str
+  for (size_t i = 0; i < cust_str_len; ++i) {
+    prefix_buf_ptr[write_cnt + i] = cust_str[i];
+  }
+  write_cnt += cust_str_len;
+
+  // Fill the remaining bytes with 0x0
+  for (size_t i = write_cnt; i < 4 * ARRAYSIZE(prefix_buffer); ++i) {
+    prefix_buf_ptr[i] = 0x0;
+  }
```

Comment:
nit: I think you don't need this, since you declared `prefix_buffer` to be pre-filled with zeroes initially.